### PR TITLE
feat: create invite

### PR DESCRIPTION
**Pull request by Micah-Shallom**
## Description
Implemented an API endpoint to generate invitation links for inviting users to join an organization. This endpoint validates the input data, ensures proper authorization, and generates a unique invitation link for the specified organization. The invitation link is stored in the database with an expiration date, and optionally, an email containing the invitation link is sent to the specified email address.
​
## Related Issue (Link to Github issue)

https://github.com/hngprojects/hng_boilerplate_golang_web/issues/108#issue-2424090317
​
## How Has This Been Tested?

The changes have been tested using the Go testing framework. The test environment was set up with Gin in test mode, mock data was created, and test cases for the CreateInvite function were executed. These tests ensured the validation of inputs, user authorization checks, and correct generation of invitation links. Various scenarios were tested, including successful invitation creation, invalid email format, missing organization ID, and unauthorized user.
​
## Screenshots 
hng

a2

a3

a4

a5

hng

​
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
​
## Checklist:

- [X] My code follows the code style of this project.
- [X] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

**Review by Cyberguru1 — DISMISSED** [review]
all good

File: internal/models/organisation.go [hunk 1]
```diff
@@ -46,3 +46,12 @@ func (c *Organisation) CreateOrganisation(db *gorm.DB) error {
 
 	return nil
 }
+
+func (o *Organisation) GetOrgByID(db *gorm.DB, orgID string) (Organisation, error) {
+	var org Organisation
+
+	if err := db.Where("id = ?", orgID).First(&org).Error; err != nil {
+		return org, err
+	}
```
Comment (NwokoyeChigozie) [inline internal/models/organisation.go]: stop stressing me nah, just follow instructions

File: pkg/controller/invite/postInvite.go [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+		// domain := config.Config.Mail.Domain
+		// apikey := config.Config.Mail.APIKey
+
+		// inviteErr := send_invites.MockSendInvite(domain, apikey, senderEmail, to, subject)
+		// if inviteErr != nil {
+		// 	// Log error and skip user
+		// 	base.Logger.Error("error sending invitation", err)
+		// 	continue
+		// }
+
```
Comment (NwokoyeChigozie) [inline pkg/controller/invite/postInvite.go]: clean up comments

File: tests/base.go [hunk 3]
```diff
@@ -105,3 +124,59 @@ func GetLoginToken(t *testing.T, r *gin.Engine, auth auth.Controller, loginData
...
+// 	req.Header.Set("Content-Type", "application/json")
+// 	req.Header.Set("Authorization", "Bearer "+token)
+
+// 	rr := httptest.NewRecorder()
+// 	r.ServeHTTP(rr, req)
+
+// 	//get the response
+// 	data := ParseResponse(rr)
+// 	return data
+// }
```
Comment (NwokoyeChigozie) [inline tests/base.go]: clean up comments

File: tests/test_invites/invite_test.go [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+// 					AssertResponseMessage(t, message.(string), test.Message)
+// 				} else {
+// 					AssertResponseMessage(t, "", test.Message)
+// 				}
+
+// 			}
+
+// 		})
+// 	}
+// }
```
Comment (NwokoyeChigozie) [inline tests/test_invites/invite_test.go]: please clean up comments

**Review by NwokoyeChigozie — CHANGES_REQUESTED** [review]

File: tests/test_invites/invite_test.go (hunk 4 above)
Comment (Micah-Shallom) [reply tests/test_invites/invite_test.go]: All highlighted issues has been corrected

File: tests/test_invites/invite_test.go [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+				}
+
+			}
+
+		})
+
+	}
+}
+
+// func TestPostInvite(t *testing.T) {
```
Comment (buka4rill) [inline tests/test_invites/invite_test.go]: Remove all these irrelevant comments

File: tests/test_invites/invite_test.go [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+					tst.AssertResponseMessage(t, "", test.Message)
+				}
+
+			}
+
+		})
+
+	}
+}
+
```
Comment (buka4rill) [inline tests/test_invites/invite_test.go]: Remove this irrelevant comment

File: tests/test_invites/invite_test.go [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+			},
+		},
+		{
+			Name: "Empty Email Field",
+			RequestBody: models.InvitationCreateReq{
+				OrganisationID: org_id,
+				Email:          "",
+			},
+			ExpectedCode: http.StatusBadRequest,
+			// Message:      "Failed to parse request body",
```
Comment (buka4rill) [inline tests/test_invites/invite_test.go]: Remove this irrelevant comment

**Review by buka4rill — COMMENTED** [review]
Take away spaghetti comments

**Comment by Micah-Shallom** [comment]
@buka4rill @NwokoyeChigozie fixed issues

File: services/invite/checker.go [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+		return org, err
+	}
+
+	if !isMember {
+		rd := utility.BuildErrorResponse(http.StatusNotFound, "error", "User not a member of the organization", err, nil)
+		c.JSON(http.StatusNotFound, rd)
+		return org, errors.New("User not a member of the organization")
+	}
+
+	return org, nil
```
Comment (NwokoyeChigozie) [inline services/invite/checker.go:93]: service function should not be doing c.JSON for any reason

File: services/invite/checker.go [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+				"expires_at":   invitation.ExpiresAt.Format(time.RFC3339),
+			},
+		)
+
+		logger.Info("Invitations posted successfully")
+		rd := utility.BuildSuccessResponse(http.StatusCreated, "Invitation(s) sent successfully", invitations)
+
+		c.JSON(http.StatusCreated, rd)
+	}
+}
```
Comment (NwokoyeChigozie) [inline services/invite/checker.go:163]: service function should not be doing c.JSON for any reason

File: services/invite/generator.go [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+	err = SaveInvitation(base.Postgresql, userId, token, inviteReq)
+	if err != nil {
+		rd := utility.BuildErrorResponse(http.StatusInternalServerError, "error", "Failed to save invitation", err, nil)
+		c.JSON(http.StatusInternalServerError, rd)
+		return "", err
+	}
+
+	invitationLink := GenerateInvitationLink("http://localhost:8019", token)
+
+	return invitationLink, nil
```
Comment (NwokoyeChigozie) [inline services/invite/generator.go:29]: service function should not be doing c.JSON for any reason

**Review by NwokoyeChigozie — CHANGES_REQUESTED** [review]

**Pull request merged**